node/router/src/helpers/resolver.rs

```diff
@@ -13,30 +13,87 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use snarkvm::prelude::{Address, Network};
+
 use std::{collections::HashMap, net::SocketAddr};
 
-/// The `Resolver` provides the means to map the connected address (used in the lower-level
-/// `tcp` module internals, and provided by the OS) to the listener address (used as the
-/// unique peer identifier in the higher-level functions).
-#[derive(Debug, Default)]
-pub(crate) struct Resolver {
-    /// The map of the connected peer address to the correponding listener address.
+/// The resolver contains some reverse maps for peers which are not available
+/// by default to the implementors of PeerPoolHandling (who already contain
+/// maps from the peer's listening address to their various components).
+#[derive(Debug)]
+pub struct Resolver<N: Network> {
+    /// The map of the (ambiguous) peer address to listener address.
     to_listener: HashMap<SocketAddr, SocketAddr>,
+    /// A map of `address` to `peer IP`.
+    address_peers: HashMap<Address<N>, SocketAddr>,
+}
+
+impl<N: Network> Default for Resolver<N> {
+    /// Initializes a new instance of the resolver.
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<N: Network> Resolver<N> {
+    /// Initializes a new instance of the resolver.
+    pub fn new() -> Self {
+        Self { to_listener: Default::default(), address_peers: Default::default() }
+    }
 }
 
-impl Resolver {
-    /// Returns the listener address for the given connected address, if it exists.
-    pub fn get_listener(&self, connected_addr: &SocketAddr) -> Option<SocketAddr> {
-        self.to_listener.get(connected_addr).copied()
+impl<N: Network> Resolver<N> {
+    /// Returns the listener address for the given (ambiguous) peer address, if it exists.
+    pub fn get_listener(&self, peer_addr: SocketAddr) -> Option<SocketAddr> {
+        self.to_listener.get(&peer_addr).copied()
     }
 
-    /// Inserts a new mapping of a connected address to the corresponding listener address.
-    pub fn insert_peer(&mut self, listener_addr: SocketAddr, connected_addr: SocketAddr) {
-        self.to_listener.insert(connected_addr, listener_addr);
+    /// Returns the peer IP for the given address.
```

Comment:
```suggestion
    /// Returns the peer IP for the given Aleo address.
```
To reduce the `SocketAddr`/`Address` confusion a little.

Also, the same comment as above. If this is the listening address, we should change it to 
```suggestion
    /// Returns the listening address for the peer with the given Aleo address.
```
(This also uses "address" twice, but it is a little clearer)